md/fire.py

```diff
@@ -0,0 +1,147 @@
+@dataclasses.dataclass
+class FireDescentState:
+    """A dataclass containing state information for the Fire Descent minimizer.
+    Attributes:
+      position: The current position of particles. An ndarray of floats
+        with shape [n, spatial_dimension].
+      velocity: The current velocity of particles. An ndarray of floats
+        with shape [n, spatial_dimension].
+      force: The current force on particles. An ndarray of floats
+        with shape [n, spatial_dimension].
+      dt: A float specifying the current step size.
+      alpha: A float specifying the current momentum.
+      n_pos: The number of steps in the right direction, so far.
+    """
+
+    position: jnp.array
+    velocity: jnp.array
+    force: jnp.array
+    dt: float
+    alpha: float
+    n_pos: int
+
+
+def fire_descent(
+    force,
+    shift_fn,
+    dt_start: float = 1e-5,
+    dt_max: float = 1e-3,
+    n_min: float = 5,
+    f_inc: float = 1.1,
+    f_dec: float = 0.5,
+    alpha_start: float = 0.1,
+    f_alpha: float = 0.99,
+):
+    """Defines FIRE minimization.
+    This code implements the "Fast Inertial Relaxation Engine" from [1].
+    Args:
+      force: A function that produces a force from
+        a set of particle positions specified as an ndarray of shape
+        [n, spatial_dimension].
+      shift_fn: A function that displaces positions, R, by an amount dR. Both R
+        and dR should be ndarrays of shape [n, spatial_dimension].
+      quant: Either a quantity.Energy or a quantity.Force specifying whether
+        energy_or_force is an energy or force respectively.
+      dt_start: The initial step size during minimization as a float.
+      dt_max: The maximum step size during minimization as a float.
+      n_min: An integer specifying the minimum number of steps moving in the
+        correct direction before dt and f_alpha should be updated.
+      f_inc: A float specifying the fractional rate by which the step size
+        should be increased.
+      f_dec: A float specifying the fractional rate by which the step size
+        should be decreased.
+      alpha_start: A float specifying the initial momentum.
+      f_alpha: A float specifying the fractional change in momentum.
+    Returns:
+      See above.
+    [1] Bitzek, Erik, Pekka Koskinen, Franz Gahler, Michael Moseler,
+        and Peter Gumbsch. "Structural relaxation made simple."
+        Physical review letters 97, no. 17 (2006): 170201.
+    """
+
+    dt_start, dt_max, n_min, f_inc, f_dec, alpha_start, f_alpha = static_cast(
+        dt_start, dt_max, n_min, f_inc, f_dec, alpha_start, f_alpha
+    )
+
+    def init_fn(R: np.array, **kwargs) -> FireDescentState:
+        V = jnp.zeros_like(R)
+        n_pos = jnp.zeros((), jnp.int32)
+        F = force(R, **kwargs)
+        return FireDescentState(
+            R, V, F, dt_start, alpha_start, n_pos
+        )  # pytype: disable=wrong-arg-count
+
+    def apply_fn(state: FireDescentState, **kwargs) -> FireDescentState:
```

Comment:
Note that the LAMMPS paper users a modified algorithm which seems to be built around a different integrator: https://arxiv.org/pdf/1908.02038.pdf

Reply:
An advantage of FIRE over scipy optimize (at least L-BFGS) is that it is faster. Definitely worth a comparison in the future.